[baremetal] Harden tinycrypt against fault injection

mpg opened the pull request:
## Description

**Note to gatekeeper:** this PR makes changes to the TinyCrypt API - see "Migrations" below.

~Dependencies:This PR is based #2918 (merged).~

This PR adds counter-measures against fault injection in ECC computations to our version of TinyCrypt. It also performs some refactorings to simplify implementation of the counter-measures and reduce their impact on code size.

The counter-measures to implement are

- [x] check that the point lies on the curve at the end of the computation
- [x] add integrity check for curve parameters (p, n, G, b)
- [x] add reporting of detected faults
- [x] add double-checking, high Hamming distance and other safety measures where necessary
- [x] document API changes

### Code size impact

Size of the library (sum of all 3 `.a` files, with SSL server enabled) in bytes, as measured with `scripts/baremetal.sh --rom --armc5 | grep total`:

| step | size | diff (prev.) | diff (first) |
| --- | --- | --- | --- |
| Before | 39842 | +0 | +0 |
| After (centralized) point validation | 39746 | -96 | -96 |
| After refactoring curve parameters | 39310 | -436 | -532 |
| After integrity check for curve parameters | 39586 | +276 | -256 |
| After reporting of detected faults | 39682 | +96 | -160 |
| After double-check / flow control | 39774 | +92 | -68 |

Thanks to the refactorings, the PR **saves 68 bytes** of code size despite adding new security checks.

### Impact on speed

The execution speed of various operations was measured on a  `NUCLEO_L073RZ` using Jarno's [test program](https://github.com/ARMmbed/mbedtls-test/tree/master/benchmark/rng_test). The difference, if any, is smaller (0 to 8 ms on a 1.9 second operation) that the measurement noise (up to 10 ms).

The counter-measures added by this PR have **no measurable impact** on speed.

## Status
**READY**

## Migrations

This PR contains API changes, which will require changing all application code using one of the affected APIs directly. The top-level ECDH and ECDSA APIs are affected, so it is very likely that all code using TinyCrypt directly will need changing. Code using TinyCrypt only through the Mbed TLS API (functions starting with `mbedtls_`) is not impacted.

The main changes are a new return value convention, and the removal of the "curve" parameter for functions that had one.

Here's a more detailed and complete list sorted by type of change.

### Change of return value convention.

Some functions used to return `1` for success and `0` for failure, and now return one of `UECC_SUCCESS`, `UECC_FAILURE`, `UECC_FAULT_DETECTED`. Failure to adapt caller code would not result in compile-time or link-time errors, but would cause the code to misbehave at runtime instead, potentially in security-critical ways. Testing might not detect all errors so code review is required.

List of public functions affected:

* `uECC_make_key()`
* `uECC_shared_secret()`
* `uECC_sign()`
* `uECC_verify()`
* `uECC_compute_public_key()`

List of semi-internal functions affected - application code should not need to call these directly unless, so hopefully no changes are needed here, but since they're technically public here's the list anyway:

* `EccPoint_compute_public_key()` - use `uECC_compute_public_key()` instead
* `EccPoint_mult_safer()` - use one of the ECDH or ECDSA APIs instead
* `EccPoint_isZero()` - use `uECC_valid_public_key()` instead

**Actions required:**

* identify all calling points of the affected function in the code
* edit all of them to check the return value as follows: (1) `UECC_SUCCESS` means success, and any other value means failure, but (2) the failing value `UECC_FAULT_DETECTED` might be treated in a special way for logging or reporting purposes, as it denotes a hardware malfunction in a security-critical path, which may be indicative of an active physical attack.

### Change of signature for some functions

The following functions lost a parameter. They were previously accepting a `curve` parameter, but since only a single value for this parameter was supported (namely the return value of `uECC_secp256r1()`), this parameter was useless and has been removed for consistency and to simplify the implementation of some internal checks.

Public functions that lost the `curve` parameter:

* `uECC_make_key()`
* `uECC_shared_secret()`
* `uECC_sign()`
* `uECC_verify()`

* `uECC_curve_private_key_size()`
* `uECC_curve_public_key_size()`
* `uECC_compute_public_key()`
* `uECC_valid_public_key()` - note: you're welcome to call this function, but shouldn't need to, as validity of public keys is already checked by all functions taking them as input anyway.

Semi-internal functions that lost the `curve` parameter - application code should not need to call these directly unless, so hopefully no changes are needed here, but since they're technically public here's the list anyway:

* `EccPoint_compute_public_key()` - use `uECC_compute_public_key()` instead
* `EccPoint_mult_safer()` - use one of the ECDH or ECDSA APIs instead
* `EccPoint_isZero()` - use `uECC_valid_public_key()` instead
* `XYcZ_add()`  - use one of the ECDH or ECDSA APIs instead
* `uECC_valid_point()` - use `uECC_valid_public_key()` instead

**Actions required:**

* edit calling sites of these functions to remove the final `curve` argument;
* if a `curve` variable was used, remove it as well as it's now unused.

### Removal of the `curve` structure and associated types and function

In relation to the previous point (functions loosing their `curve` parameter), the following parts of the API were removed:

* the types `struct uECC_Curve_t` and `uECC_Curve`
* the object `curve_secp256r1`
* the function 

**Actions required:** none in addition to what was already done to adapt to the signature change above.

### Removal of a few internal functions from the public API

The following functions were previously publicly visible while being internal in nature. It is very likely that no application code was using them, but here's the list, with suggestions for replacement:

* `x_side_default()`: use `uECC_valid_public_key()` instead

**Actions required:** hopefully none as you're probably not calling those functions already. Otherwise, adapt code to make use of the suggested higher-level function instead.

mpg commented:
The pr-merge job succeeded, and the pr-head job only failed in one instance of the known-flaky "reconnect" `ssl-opt.sh` test - restarting it.

artokin commented on tinycrypt/ecc.c:
```diff
@@ -98,6 +99,73 @@ const uECC_word_t curve_b[NUM_ECC_WORDS] = {
+static int uECC_check_curve_integrity(void)
+{
+	unsigned char computed[32];
+	unsigned char reference[32] = {
```
Can `reference` be const table?

mpg replied:
OMG yes. Should probably be static as well.

artokin commented on tinycrypt/ecc_dsa.c:
```diff
@@ -159,16 +159,17 @@ int uECC_sign_with_k(const uint8_t *private_key, const uint8_t *message_hash,
 	      unsigned hash_size, uint8_t *signature)
 {
+	int r;
 	      uECC_word_t _random[2*NUM_ECC_WORDS];
```
Would you please fix the function indentation?

mpg replied:
Sure. TinyCrypt indentation is often broken so I don't always fix it, but here since I'm making and edit in the same paragraph it makes sense indeed.

mpg commented:
The pr-merge job fully passed, but the pr-head job has two failures that both look spurious - restarting.

mpg commented:
@artokin Thanks for your review. I think I've addressed your feedback and this is now ready for you to review again.

@jarlamsa Please also review this when you get the chance. Thanks!

jarlamsa commented on tinycrypt/ecc.c:
```diff
@@ -1052,17 +1052,17 @@ int EccPoint_mult_safer(uECC_word_t * result, const uECC_word_t * point,
 	}
 
-	r = 1;
+	r = UECC_SUCCESS;
```
Here we could do some reference counting that each step has been completed succesfully, potential attacker could glitch out one of the branches before setting this to true.

jarlamsa replied:
The issue here was addressed by doing double checks, LGTM

mpg commented:
With two approvals and a passing CI, no conflicts, and the only prerequisite also ready for merge, this is ready for merge.

jarlamsa commented on tinycrypt/ecc.c:
```diff
@@ -1037,82 +1167,88 @@ int uECC_generate_random_int(uECC_word_t *random, const uECC_word_t *top,
 	uECC_word_t tmp2[NUM_ECC_WORDS];
-	wordcount_t num_words = curve->num_words;
+	wordcount_t num_words = NUM_ECC_WORDS;
+	volatile uECC_word_t diff = -1u;
```
VS 2013 doesn't like this line
`warning C4146: unary minus operator applied to unsigned type, result still unsigned`

jarlamsa commented:
@jarvte could you review the latest commit to this PR, @artokin, please re-review.

jarlamsa commented:
With two approvals and a passing CI, no conflicts, and the only prerequisite also ready for merge, this is ready for merge.

sbutcher-arm commented:
Failing IAR compiler with: 
```
IAR ANSI C/C++ Compiler V8.20.1.14183/W32 for ARM
   Copyright 1999-2017 IAR Systems AB.
   Network license: 10.10.12.86 (STD)

          diff |= computed[i] ^ reference[i];
          ^
"C:\builds\ws\mbedtls-release-new\src\tinycrypt\ecc.c",163  Error[Pa082]: 
          undefined behavior: the order of volatile accesses is undefined in
          this statement

      diff |= (unsigned char) i ^ 32;
      ^
"C:\builds\ws\mbedtls-release-new\src\tinycrypt\ecc.c",167  Error[Pa082]: 
          undefined behavior: the order of volatile accesses is undefined in
          this statement

          diff |= (left[i] ^ right[i]);
          ^
"C:\builds\ws\mbedtls-release-new\src\tinycrypt\ecc.c",287  Error[Pa082]: 
          undefined behavior: the order of volatile accesses is undefined in
          this statement

Errors: 3
```
Therefore marking as 'needs: work' and removing 'ready to merge'.

mpg commented:
@sbutcher-arm Thanks for reporting the error! I hopefully fixed it in a new commit.

@artokin @jarlamsa Could you review the latest commit? Thanks!

mpg commented:
The base PR #2918 has been merged, so removing "needs: preceding PR".

mpg commented:
(Changed the base twice in order to flush github's cache and have a correct commit list.)

mpg commented:
Applying "DO NOT MERGE" and "needs: work" as the PR contains API changes which are not documented so far.

mpg commented:
Removing "DO NOT MERGE" and "needs: work" as API changes are now properly documented in the PR description.

The CI is only failing the API/ABI check, which is expected and now documented in the PR description, so removing "needs: ci" as well.

The PR is now only awaiting one more re-review - @jarlamsa or @jarvte could one of you review the commit added since your previous approval? Thanks!

mpg commented:
This PR has two approvals, and no conflicts or dependencies.

The CI is only failing the API/ABI check, which is expected as the API is indeed changed. The changes are fully documented in the PR description. All other tests are passing.

So labeling "ready for merge".
